### ISSUE TITLE
Handling of composite types in type checker

### ISSUE THREAD

**Issue by nlmave** [overturetool/overture#461]
class CheckType

    types

    -- I cannot declare atype as public because ctype is implicity private
    -- this is (probably) OK albeit a quirky effect of the compose construct
    -- we should probably annotate this (better) in the LRM, or consider
    -- (1) to extend the syntax to allow 'compose public ctype of' -or-
    -- (2) discard the 'compose of' construct as it does not bring anything
    -- but that is not the main point of this report, see further on
    atype =
    		compose ctype of
    			a : nat
    		end;
		
    -- note that using the short-hand notation it is possible to change
    -- the visibility; which actually gives us more flexibility than compose-of
    public ctype :: a : nat;

    -- however, it exposes another problem (imho a bug) because the type checker
    -- does allow the type name ctype to be redefined without raising an error
    -- which leads to a "type component visibility" error in dtype below as the
    -- look-up probably refers to the implicitly private compose-of ctype rather
    -- than the redefined public type
    public dtype = ctype
				
    end CheckType

**Comment by nlmave** [comment]
Interestingly, if I change the second definition to:

    ctype :: a : nat b: real;

then a type error is generated that will catch the redefinition

**Comment by nickbattle** [comment]
VDMJ is giving me a duplicate type definition on the original and modified specs. I haven't tried on Overture yet:
```
Error 3017: Duplicate definitions for ctype in 'CheckType' (test.vpp) at line 12:17
ctype: in 'CheckType' (test.vpp) at line 12:17
ctype: in 'CheckType' (test.vpp) at line 18:8
Error 3321: Type component visibility less than type's definition in 'CheckType' (test.vpp) at line 25:8
```
With your modification, I get a mismatched compose definition as well:
```
Error 3017: Duplicate definitions for ctype in 'CheckType' (test.vpp) at line 12:17
ctype: in 'CheckType' (test.vpp) at line 12:17
ctype: in 'CheckType' (test.vpp) at line 19:1
Error 3325: Mismatched compose definitions for ctype in 'CheckType' (test.vpp) at line 12:17
ctype: in 'CheckType' (test.vpp) at line 12:17
ctype: in 'CheckType' (test.vpp) at line 19:1
Error 3321: Type component visibility less than type's definition in 'CheckType' (test.vpp) at line 27:8
```
As you say, that suggests that the final use of ctype is picking up the compose one. But since there is a duplicate error as well, that might be "ok". I'll try on Overture when I get a minute.

**Comment by nickbattle** [comment]
Confirmed that using Overture-2.2.6.jar, the duplicate definition check does not occur with the first specification, but it does with the second. Curious. I'll step through it.

**Comment by nickbattle** [comment]
OK, this is now fixed in ncb/development. There was a line missing where the list of defined types is collected that neglected to include the auto-generated compose type. Correcting that now gives the same errors as VDMJ.

**Comment by nickbattle** [comment]
I agree that the "private" implicit compose type isn't very helpful. I would suggest we change it so that any generated types have the same access specifier as the type generating them - here, that would still be private, but changing atype to be public would create a public ctype and solve the problem.

Does that sound OK? I'd rather not monkey with the syntax, or prohibit compose types, as several older specs use them.

**Comment by nlmave** [comment]
I'm still somewhat puzzled by the 3325 error: it seems to hint at the possibility to redefine the type? I would assume that this error is void because duplicate definitions are simply not allowed? Can you give an example of a situation where 3325 is "normally" used?

And I do not fully understand the solution you propose to address the implicit private compose-of, can you provide a concrete example how you want to solve it?

**Comment by nickbattle** [comment]
Error 3325 is the result of a bug we had a while ago, where there could be clashes between the explicit definition of a type and the implicit definition of the same type via an embedded "compose". So now, when you mention (and therefore implicitly create) a type with compose, the type checker makes sure that any existing record definition of the same name has the same field definitions. Since your second (explicit) definition of ctype is not the same (it has a b:real) as the implicit one inside atype, you get the error. If the two types had the same signature, there would be no error.

Regarding the public/private access, I was suggesting that "private T = compose R of..." would create an implicit private R, "public T = compose R of..." would create a public R, and similarly for protected. Obviously omitting the specifier is the same as using private. It's slightly odd, but perhaps better than always making the implicit types private?

So we now need to be clear what happens if a private implicit R field-matches a public explicit R. Is the type public or private, or is this an error? (I suspect the explicit type access wins currently).

**Comment by nlmave** [comment]
OK I now better understand the 3325 error; we insist on structural equivalence in case the type name is used both inside the compose-of and as another explicitly defined type.

Taking that philosophy one step further, it should probably also hold for the type visibility, leading to an error in case the compose-of version deviates from the explicit type with the same name.

I propose we close this error and submit a new RM to the LB with the proposal you outline above + fix the LRM to better describe this behavior of compose-of i.r.t. type visibility in VDM++.

**Comment by nickbattle** [comment]
In the past, we've handled these sorts of "clean up" issues with bugs/fixes rather than the full blown LB RM process. If there is no disagreement and the solution is relatively obvious, then this is much simpler - though I agree we have to make sure the LRM is updated somehow.

So perhaps we just create a new issue? The fix should be trivial.

**Comment by nlmave** [comment]
Ok I will take care to update the LRM
Op 28 jul. 2015 19:16 schreef Nick Battle <notifications@github.com>:In the past, we've handled these sorts of "clean up" issues with bugs/fixes rather than the full blown LB RM process. If there is no disagreement and the solution is relatively obvious, then this is much simpler - though I agree we have to make sure the LRM is updated somehow.

So perhaps we just create a new issue? The fix should be trivial.

**Comment by nickbattle** [comment]
Looking at this last night, I think I was confused about how we intended this to work, after the fix to #243. The change only implicitly defines new record types when the compose types appear as part of a normal "T =" type definition. Other *uses* of compose merely reference the type, and it is those references that must match. So it makes sense to say that the implicit record type should inherit the same access specifier as the "T" definition. But if two type definitions define the same implicit record type (whether the same shape or not), this is a duplicate record definition error. So we don't have to additionally check the access specifiers - they are duplicates, and so an error. There is no "access specifier" for the *usage* of compose records elsewhere, so that isn't an issue either.
